File: 백준/Gold/7490. 0 만들기/0 만들기.java
import java.io.*;
import java.util.ArrayList;
import java.util.Collections;
import java.util.StringTokenizer;

public class Main {
    public static ArrayList<String> result;

    public static int parse(String expr) {
        expr = expr.replaceAll(" ", "");
        ArrayList<Integer> numbers = new ArrayList<>();
        ArrayList<Character> op = new ArrayList<>();

        StringBuilder num = new StringBuilder();

        for (int i = 0; i < expr.length(); i++) {
            char c = expr.charAt(i);

            if (c == '+' || c == '-') {
                op.add(c);
                numbers.add(Integer.parseInt(num.toString()));
                num = new StringBuilder();
            } else {
                num.append(c);
            }
        }

        if (num.length() > 0) {
            numbers.add(Integer.parseInt(num.toString()));
        }

        int cur = 1;
        int total = numbers.get(0);

        for (char c : op) {
            if (c == '+') {
                total += numbers.get(cur);
            } else {
                total -= numbers.get(cur);
            }
            cur += 1;
        }

        return total;
    }

    public static void dfs(String expr, String num, int cnt) {
        char[] operation = new char[] {'+', '-', ' '};

        if (cnt == num.length()) {
            if (parse(expr) == 0) {
                result.add(expr);
            }
            return;
        }

        for (char c : operation) {
            dfs(expr + c + num.charAt(cnt), num, cnt + 1);
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int t = Integer.parseInt(st.nextToken());

        for (int i = 0; i < t; i++) {
            StringBuilder word = new StringBuilder();
            result = new ArrayList<>();
            st = new StringTokenizer(br.readLine());
            int n = Integer.parseInt(st.nextToken());

            for (int j = 1; j <= n; j++) {
                word.append(j);
            }

            dfs("1", word.toString(), 1);

            Collections.sort(result);
            for (String expr : result) {
                System.out.println(expr);
            }

            System.out.println();
        }
    }
}

File: 백준/Gold/7490. 0 만들기/0 만들기.py
# 0만들기
import sys
input = sys.stdin.readline

def parse(expr):
    expr = expr.replace(" ", "")
    numbers = []
    op = []
    
    num = ""
    for i in range(len(expr)):
        if expr[i] == "+" or expr[i] == "-":
            op.append(expr[i])
            numbers.append(int(num))
            num = ""
        else:
            num += expr[i]

    if len(num) > 0:
        numbers.append(int(num))

    cur = 1
    total = numbers[0]

    for i in op:
        if i == "+":
            total += numbers[cur]
        else:
            total -= numbers[cur]
        cur += 1
    
    return total

def dfs(expr, num, cnt):
    oper = ("+", "-", " ")
    
    if cnt == len(num):
        if parse(expr) == 0:
            result.append(expr)
        return

    if cnt == len(num):
        return
    
    for i in oper:
        dfs(expr+i+num[cnt], num, cnt+1)

t = int(input())

for i in range(t):
    num = int(input())
    word = ""
    result = []
    
    for j in range(1, num+1):
        word += str(j)

    dfs(word[0], word, 1)
    result.sort()
    
    for i in result:
        print(i)
    print()
    
